Proper List formatting

Dark-SM opened the pull request:
_(It's in draft form for now since I'm not even halfway done. Maybe a simple script can take care of this easier)_

I made this pull request to improve how lists are formatted by Alpaca.

**Input:** `Generate a list of random words.`

**Before:**

`sneeze, conflict, ancestor, thunder, companion, amulet, forge, serpent, magnify, jester, cascade, universe, surge, nausea, oasis, anomaly, pursue, rapture, pagoda, mosaic.`

**After:**
- sneeze
- conflict
- ancestor
- thunder
...

I hope others want to join in and help improve Alpaca's formatting.

Dark-SM committed:
First Commit

Dark-SM committed:
making progress

HideLord commented:
Here's a little script to help with your endeavor.
```
import re
import json

# Can be played with.
MIN_NUMBER_OF_ITEMS_IN_LIST = 4
# There are quite a few questions of type: 'Sort the following list in alphabetical order.' We usually don't want to format those items.
CHECK_IF_SORT_QUESTION = True
# Check if there is a number in the instruction. Using numbered lists helps the LLM correctly answer queries that want a specific number of answers.
CREATE_NUMBERED_LIST_IF_NUMBER_IN_INSTRUCTION = True
NUM_MAP = {'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5, 'six': 6, 'seven': 7, 'eight': 8, 'nine': 9, 'ten': 10}

with open('alpaca_data_cleaned.json', 'r') as f:
	data = json.load(f)

pattern = re.compile(r'^((?:(?:[a-zA-Z]+[ \-]){0,2}[a-zA-Z]+, ){' + str(MIN_NUMBER_OF_ITEMS_IN_LIST-1) + r',}(?:and |or )?(?:[a-zA-Z]+[ \-]){0,2}[a-zA-Z]+)\.?$')

num_matches = 0
for item in data:
	if CHECK_IF_SORT_QUESTION and 'sort' in item['instruction'].lower():
		continue
	
	match = re.search(pattern, item['output'])

	if match:
		num_matches += 1
		
		comma_sep_list = match.group(1)
		md_list = re.sub(r', (and |or )?', '\n- ', comma_sep_list)
		md_list = '\n'.join(['- ' + x.strip().title() for x in md_list.split('\n- ')])
		
		if CREATE_NUMBERED_LIST_IF_NUMBER_IN_INSTRUCTION:
			instruction = item['instruction']
			num_match = re.search(r'\b(?:\d+|one|two|three|four|five|six|seven|eight|nine|ten)\b', instruction.lower())
			if num_match:
				num_items = num_match.group(0)
				if num_items.isdigit():
					num_items = int(num_items)
				else:
					num_items = NUM_MAP.get(num_match.group(0), -1)
				md_list_items = md_list[2:].split('\n- ')
				if len(md_list_items) < num_items:
					print(f"Error: {item['output']} has {len(md_list_items)} items, but the instruction asks for {num_items} items.")
				md_list = '\n'.join([f'{i+1}. {x.title()}' for i, x in enumerate(md_list_items)])

		item['output'] = item['output'].replace(match.group(0), md_list)

print(f'Matched {num_matches} patterns.')

with open('alpaca_new.json', 'w') as f:
	json.dump(data, f, indent=4)
```

It relies on the oxford comma when there is `and` or `or`, so it might break if there isn't one, but it should generally work for 99% of cases.

E.G:
```
"instruction": "Provide five adjectives that describe a particular animal.",
"input": "Lion",
"output": "Regal, mighty, majestic, brave, intelligent."
```
1. Regal
2. Mighty
3. Majestic
4. Brave
5. Intelligent

Dark-SM committed:
Json modified with HideLord's code (tysm), NEEDS REVIEWING, STILL NOT READY

Dark-SM commented:
This code did wonders, should have worked smart, not hard. Not only did it do things I missed (such as: '_Provide five_' or '_Generate 10_' and etc.), but the error code it initially made also helped me realise that I had made a mistake in one of my commits.
( _I put \ before a number, imagine if that was left in xd_ )

Still might need some reviewing to make sure there's no mistakes, but yeah, thank you so much!

Dark-SM committed:
fixes to the script-modified json, might need to change the code to account for these idk

Dark-SM committed:
Revert "making progress"

This reverts commit f9490170253467b1f6b7c1b72c151ce20773daa3.

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -12177,7 +12177,7 @@
         "instruction": "Add three words to the following list to make it a list of animals",
         "input": "Eagle, Cat",
-        "output": "Eagle, Cat, Bear, Lion, Fox."
+        "output": "Eagle, Cat\n1. Bear\n2. Lion\n3. Fox"
```
Should the output be a complete list?

gururise replied:
The change should be:
```
"input": "1. Eagle\n2.Cat",
"output": "1. Eagle\n2. Cat\n3. Bear\n4. Lion\n5. Fox"
```

The reasoning is such an input should prime the model to generate a list output.

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -12177,7 +12177,7 @@
     {
         "instruction": "Add three words to the following list to make it a list of animals",
         "input": "Eagle, Cat",
```
Maybe the input could be a list as well so it could learn to deal with list inputs?
`"1. Eagle\n2. Cat"`

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -13282,7 +13282,7 @@
         "instruction": "Add two more animals to this list",
         "input": "Tiger, lion, wolf",
-        "output": "Tiger, lion, wolf, cheetah, elephant"
+        "output": "tiger, lion, wolf\n1. cheetah\n2. elephant"
```
Output could be a complete list:  
`"output": "1. tiger\n2. lion\n3. wolf\n4. cheetah\n5. elephant"`

gururise replied:
```
"input": "1. Tiger\n2. lion\n3. wolf",
"output": "1. Tiger\n2. lion\n3. wolf\n4. cheetah\n5. elephant"
```

Dark-SM committed:
Changed the script, plus manual labor

Dark-SM commented:
I am basically done thanks to some modification to the script which accounts for things that are not supposed to be reformatted.
Now I have to handle these pesky conflicts.....

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -30957,7 +30957,7 @@
     {
         "instruction": "Identify the musical instruments in the given song.",
         "input": "(link to a song)",
```
Model shouldnt be accepting links?

gururise replied:
should probably delete this instruction completely.

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -21722,7 +21722,7 @@
         "instruction": "Create a computer program to output all the possible combinations that can be made from the following letters: GDHAC",
         "input": "",
-        "output": "G, D, H, A, C, GD, GH, GA, GC, DH, DA, DC, HA, HC, AC, GHD, GHA, GHC, GDA, GDC, GHA, GHC, GAC, DHA, DHC, DAC, HDA, HDC, HAC, DAC."
+        "output": "- G\n- D\n- H\n- A\n- C\n- Gd\n- Gh\n- Ga\n- Gc\n- Dh\n- Da\n- Dc\n- Ha\n- Hc\n- Ac\n- Ghd\n- Gha\n- Ghc\n- Gda\n- Gdc\n- Gha\n- Ghc\n- Gac\n- Dha\n- Dhc\n- Dac\n- Hda\n- Hdc\n- Hac\n- Dac"
```
Output doesn't make sense, either have to change the instruction, or update answer with a program.

gururise replied:
Possible fix:
```
"instruction": "Output a list of all the possible combinations that can be made from the following letters: GDHAC",
"input": "",
"output": "Here's a list of all the possible combinations that can be made from the letters GDHAC:

-     G
-     D
-     H
-     A
-     C
-     GD
-     GH
-     GA
-     GC
-     DH
-     DA
-     DC
-     HA
-     HC
-     AC
-     GDA
-     GDC
-     GHA
-     GHC
-     GAC
-     DHA
-     DHC
-     DAC
-     DCA
-     HAC
-     GDAH
-     GDAC
-     GDHC
-     GHAC
-     DAHC
-     GDHAC

There are a total of 27 possible combinations of these letters."
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -34602,7 +34602,7 @@
         "instruction": "Describe a specific emotion using five words.",
         "input": "",
-        "output": "Uneasy, anxious, restless, edgy, uneasy."
+        "output": "1. Uneasy\n2. Anxious\n3. Restless\n4. Edgy\n5. Uneasy"
```
Duplicate word here

gururise replied:
ChatGPT gives the following answer to the instruction:
```
"output": "\"Melancholy\" - Sadness tinged with nostalgia."
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -35277,7 +35277,7 @@
     {
         "instruction": "Given the following problem, find the shortest path and output the nodes.",
```
this problem doesn't really make sense

gururise replied:
```
"output": "I'm sorry, but you have not provided enough information about the problem to determine the shortest path.\nPlease provide more context or details about the problem such as the nodes and their connections or weights between them. With that information, I will be able to help you find the shortest path and output the nodes."
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -36662,7 +36662,7 @@
         "instruction": "Edit the given list so that all items are gramatically correct and consistent.",
         "input": "Olive, bananaes, potato, eggs",
-        "output": "Olives, bananas, potatoes, eggs"
+        "output": "- Olives\n- Bananas\n- Potatoes\n- Eggs"
```
Maybe for editing given lists, the output shouldn't change the format of the input?

gururise replied:
Agreed. If the input is not in numbered or bulleted list format, and if the instruction doesn't say to arrange the output into a list (or bulleted list), we should probably keep the same format.

minh-v reviewed (commented):
Did a initial quick look through, have some feedback

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -57697,7 +57697,7 @@
         "instruction": "Order the following animals according to their size.",
         "input": "Marmot, Elephant, Wolf, Gorilla",
-        "output": "Marmot, Wolf, Gorilla, Elephant"
+        "output": "- Marmot\n- Wolf\n- Gorilla\n- Elephant"
```
Maybe this shouldn't change the list format for this instruction?

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -59957,7 +59957,7 @@
         "instruction": "Find the circumference of a circle of radius R",
         "input": "5",
-        "output": "2Rπ = 10π ≈ 31.4159"
+        "output": "2R\u03c0 = 10\u03c0 \u2248 31.4159"
```
can add CoT here

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -60187,7 +60187,7 @@
     {
         "instruction": "Find five words that have the definition of cheerful.",
         "input": "<ninput>",
```
`"<ninput>"` here

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -70417,7 +70417,7 @@
         "instruction": "Identify the correct sequence",
         "input": "A, B, C, D",
-        "output": "A, B, C, D"
+        "output": "- A\n- B\n- C\n- D"
```
This instruction is confusing

gururise replied:
```
"output": "I'm sorry, but without further context, it is not clear what you are asking me to identify as the correct sequence."
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -72782,7 +72782,7 @@
         "instruction": "Generate alternative names for a new business.",
         "input": "",
-        "output": "Radiant Networking, Brilliant Bridge Solutions, Marvelous Markets, Brilliant Brands International, Limitless Logic, Illuminated Innovations, Splendid Streams, Inc."
+        "output": "- Radiant Networking\n- Brilliant Bridge Solutions\n- Marvelous Markets\n- Brilliant Brands International\n- Limitless Logic\n- Illuminated Innovations\n- Splendid Streams\n- Inc"
```
Remove last `\n` to be  
`"output": "- Radiant Networking\n- Brilliant Bridge Solutions\n- Marvelous Markets\n- Brilliant Brands International\n- Limitless Logic\n- Illuminated Innovations\n- Splendid Streams Inc"`

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -77122,7 +77122,7 @@
         "instruction": "Find root mean square (RMS) of the following numbers.",
         "input": "3, 5, 8",
-        "output": "Root mean square = 7*sqrt(2/3)≈5.71548"
+        "output": "Root mean square = 7*sqrt(2/3)\u22485.71548"
```
Do we want to add CoT for this?

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -78112,7 +78112,7 @@
         "instruction": "List three items in my grocery list.",
         "input": "Apples, eggs, and milk",
-        "output": "Apples, eggs, milk, carrots, onions, and tomatoes."
+        "output": "1. Apples\n2. Eggs\n3. Milk\n4. Carrots\n5. Onions\n6. Tomatoes"
```
Weird instruction and output.

gururise replied:
what happened here was the output and input were swapped. Correct fix should be:
```
        "instruction": "List three items in my grocery list.",
        "input": "Apples, eggs, milk, carrots, onions, and tomatoes."
        "output": "1. Apples\n2. Eggs\n3. Milk"
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -88917,7 +88917,7 @@
         "instruction": "Construct a 20 word sentence using only adjectives",
         "input": "",
-        "output": "Vibrant, glowing, radiant, brilliant, shimmering, lustrous, dazzling, gleaming, shiny, glossy, luminous, sparkling, glimmering, brilliant, splendid, radiant, sublime, gorgeous, resplendent, and luminescent."
+        "output": "1. Vibrant\n2. Glowing\n3. Radiant\n4. Brilliant\n5. Shimmering\n6. Lustrous\n7. Dazzling\n8. Gleaming\n9. Shiny\n10. Glossy\n11. Luminous\n12. Sparkling\n13. Glimmering\n14. Brilliant\n15. Splendid\n16. Radiant\n17. Sublime\n18. Gorgeous\n19. Resplendent\n20. Luminescent"
```
Weird output, but sentence shouldn't be a list

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -92387,7 +92387,7 @@
         "instruction": "Analyze the text and extract a list of five nouns.",
         "input": "Once upon a time, there was a small town located deep in a valley between two mountains.",
-        "output": "town, valley, two mountains, time, small"
+        "output": "1. Town\n2. Valley\n3. Two Mountains\n4. Time\n5. Small"
```
Small is not a noun but maybe fixing answers isn't appropriate for this repo

gururise replied:
We should correct if the answer is wrong:
```
"output": "There are own four nouns in the sentence:\n\n1. Town\n2. Valley\n3. Two Mountains\n4. Time"
```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -92747,7 +92747,7 @@
         "instruction": "Generate a beat or a rhythm.",
         "input": "",
-        "output": "dum tsk, tsk tsk, dum tsk, tsk tsk, dum dum, dum dum, dum dum."
+        "output": "- Dum Tsk\n- Tsk Tsk\n- Dum Tsk\n- Tsk Tsk\n- Dum Dum\n- Dum Dum\n- Dum Dum"
```
Not sure if output and/or list is appropriate here?

gururise replied:
Should not be a list.

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -93432,7 +93432,7 @@
         "instruction": "Generate five-word sentences that provide a description of a person.",
         "input": "",
-        "output": "Energetic, determined, creative, playful, kind."
+        "output": "1. Energetic\n2. Determined\n3. Creative\n4. Playful\n5. Kind"
```
Sentence shouldn't be a list

gururise replied:
Lets use ChatGPT output:
```
1.     She had piercing blue eyes.
2.     He had a rugged beard.
3.     She had a contagious smile.
4.     He had a deep voice.
5.     She had a petite figure.

```

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -94172,7 +94172,7 @@
         "instruction": "Replace the placeholder words with a list of 8 fruits",
         "input": "<placelholder1> , <placelholder2>, <placelholder3>, <placelholder4>, <placelholder5>, <placelholder6>, <placelholder7>, <placelholder8>",
-        "output": "Apple, banana, orange, strawberry, pear, peach, mango, kiwi"
+        "output": "1. Apple\n2. Banana\n3. Orange\n4. Strawberry\n5. Pear\n6. Peach\n7. Mango\n8. Kiwi"
```
Not sure if output should change the format here

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -95392,7 +95392,7 @@
         "instruction": "Determine the length of the side of a square given the area.",
         "input": "24",
-        "output": "The length of the side of the square is sqrt(24) = 2*sqrt(6) ≈ 4.9 units."
+        "output": "The length of the side of the square is sqrt(24) = 2*sqrt(6) \u2248 4.9 units."
```
Could add CoT here

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -100772,7 +100772,7 @@
         "instruction": "Place the following countries in order of time zones from east to west.",
         "input": "Japan, Colombia, France, Australia",
-        "output": "Japan, Australia, France, Colombia"
+        "output": "- Japan\n- Australia\n- France\n- Colombia"
```
Not sure if output should change the format here

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -101477,7 +101477,7 @@
         "instruction": "Construct a sentence using a list of emotive words",
         "input": "discouraged, empty, betrayed, hopeless",
-        "output": "I felt discouraged, empty, betrayed, and hopeless."
+        "output": "- I Felt Discouraged\n- Empty\n- Betrayed\n- Hopeless"
```
Output shouldn't be changed to a list for this instruction

minh-v commented on alpaca_data_cleaned.json:
```diff
@@ -102217,7 +102217,7 @@
         "instruction": "Given some numbers, categorize the numbers into odd or even.",
         "input": "3, 7, 4, 19",
-        "output": "Odd, Odd, Even, Odd"
+        "output": "- Odd\n- Odd\n- Even\n- Odd"
```
Weird output here

gururise replied:
```
"output": "- (3) odd\n- (7) odd\n- (4) even\n- (19) odd"
```

minh-v reviewed (commented):
Few more I've found

Dark-SM commented:
Thank you so much for telling me about these. I modified the script while reading your corrections. I also fixed some of the stuff manually.

But I have one question, what's a CoT?

Dark-SM committed:
more changes to the script and other changes thanks to minh-v and gururise

matusstas commented:
CoT stands for _Chain of thought_. You can find more about this here: https://learnprompting.org/docs/intermediate/chain_of_thought

Dark-SM commented:
I don't know. I think CoT should be handled by someone who speaks and writes native English. Should I just ask ChatGPT for these?

gururise commented:
> Thank you so much for telling me about these. I modified the script while reading your corrections. I also fixed some of the stuff manually.
> 
> But I have one question, what's a CoT?

There's an example in issue #4

gururise commented:
> I don't know. I think CoT should be handled by someone who speaks and writes native English. Should I just ask ChatGPT for these?

If you are uncomfortable with adding CoT to the math-related answers, just leave CoT out, and we'll go back in and add it later.

gururise commented:
merge conflicts fixed in #40 

Thank You for the PR!
